File: 0139-word-break/0139-word-break.cpp
class Solution {
public:
    bool solve(int index, string& s, unordered_set<string>& st, vector<int>& dp){
        
        if(index >= s.size()){
            return true;
        }
        
        if(dp[index] != -1)
            return dp[index];
        
        if(st.find(s) != st.end())
            return true;
        
        for(int l=1; l<=s.size(); l++){
            string str = s.substr(index, l);
            if(st.find(str) != st.end() && solve(index+l, s, st, dp)){
                return dp[index] =  true;
            }
        }
        
        return dp[index] = false;
    }
    
    bool wordBreak(string s, vector<string>& wordDict) {
        
        unordered_set<string> st;
        for(auto& word: wordDict){
            st.insert(word);
        }
        
        vector<int> dp(s.length()+1, -1);
        return solve(0, s, st, dp);
    }
};

File: 0139-word-break/0139-word-break.java
class Solution {
    private Boolean[] dp;
    public boolean solve(int index, String s, HashSet<String> set){
        
        if(index >= s.length())
            return true;
        
        if(dp[index] != null)
            return dp[index];
        
        if(set.contains(s)==true)
            return true;
        
        for(int endIndex=index+1; endIndex<=s.length(); endIndex++){
            
            String str = s.substring(index, endIndex);
            if(set.contains(str)==true && solve(endIndex, s, set)==true){
                    return dp[index] = true;
            }
        }
        
        return dp[index] = false;
    }
    
    public boolean wordBreak(String s, List<String> wordDict) {
        
        HashSet<String> set = new HashSet<>();
        for(String word: wordDict){
            set.add(word);
        }
        
        dp = new Boolean[s.length()+1];
        
        return solve(0, s, set);
    }
}